docs/proposals/image-exec.md

```diff
@@ -0,0 +1,278 @@
+covers as create, start, attach, and remove in immediate succession.
+
+## API Changes
+
+The Exec handler will gain a new parameter to specify a debug image.
+
+TODO: details of API changes
+
+## Changes to kubelet
+
+The kubelet will gain a new `ExecInPod` method to be called by `ServeExec`
+instead of `ExecInContainer` when an container image name is provided. Image
+exec could be accommodated by the current `ExecInContainer` with the addition of
+a parameter for imageName, but it would the method name misleading.
+
+A corresponding `ExecInPod` will be added to the kubelet's container.Runtime
+interface.
+
+### Container Runtime Interface
+
+The [Container Runtime Interface](container-runtime-interface-v1.md) implements
+a standard, imperative container interface across all container runtimes
+supported by Kubernetes. This interface makes it easy to implement Image Exec.
+
+The kubelet's Generic Runtime Manager can implement `ExecInPod` with almost no
+changes to the CRI, which means we don't have to implement the logic in each of
+the container runtimes.
+
+To bind mount filesystems of other containers, the kubelet must be able to query
+the mount path of a particular container. This will be added to the `Container`
+message returned by `ListContainers`. If the container's filesystem is no longer
+mounted in the node's mount namespace then it will not be made visible to the
+troubleshooting container.
+
+`ExecInPod` will build a map of container names to node filesystem mount points
+for inclusion in the troubleshooting container's `ContainerConfig` with a
+`container_path` of `/c/<container_name>`. This can only be accomplished when
+the container is created.
+
+We could instead modify CRI's `CreateContainer` to include a list of containers
+IDs to "link" to the newly created container, but then we'd have to repeat the
+logic in each of the container backends.
+
+### Legacy Container Runtimes
+
+Implementation is slightly more difficult for the legacy docker and rkt
+runtimes.  We could implement the same create/start/attach/remove logic in rkt
+and docker_manager, or we could save effort by linking this feature to the CRI
+and return "Not Implemented" for the legacy runtimes.
+
+## Open Questions
+
+* Will Kubernetes evict the troubleshooting image for using additional
+  resources? What happens if SyncPod() is called?
+* What's the timeline for switching to the CRI?
+* Docker 1.12 leaves mounted the merged container filesystem with the aufs and
+  overlay drivers, but the current Google Cloud Image (GCI) does not. This will
+  be significantly less useful when container filesystems are inaccessible.
+* Docker 1.12 introduced a shared PID namespace. When will this be supported by
+  Kubernetes?
+* Is this compatible with Admission control plugins that restrict what images
+  can run on a cluster?
+* Would it be possible to Image Exec into pods that are in a terminal state to
+  determine why they've failed?
```

Comment:
I looked into this a little today. The current behavior in exec is that the process continues to run but there doesn't seem to be any way to reattach to it. It doesn't get a signal and stdin is left open. My prototype for image exec behaves the same.

A debug container would behave the same, except perhaps that it could be attached with `kubectl attach`, with the kubelet removing the container when the pod is destroyed or when the process exits (potentially relying on the runtime to support automatically removing the container a la "docker run --rm")